linkerd/proxy/spire-client/src/api.rs

```diff
@@ -0,0 +1,151 @@
+use futures::prelude::*;
+use linkerd_error::{Error, Recover, Result};
+use linkerd_exp_backoff::{ExponentialBackoff, ExponentialBackoffStream};
+use linkerd_identity::DerX509;
+use linkerd_identity::Id;
+use linkerd_proxy_http as http;
+use linkerd_tonic_watch::StreamWatch;
+use spiffe_proto::client::{
+    self as api, spiffe_workload_api_client::SpiffeWorkloadApiClient as Client,
+};
+use std::collections::HashMap;
+use tower::Service;
+use tracing::error;
+
+#[derive(Clone, Debug)]
+pub struct Svid {
+    pub spiffe_id: Id,
+    pub leaf: DerX509,
+    pub private_key: Vec<u8>,
+    pub intermediates: Vec<DerX509>,
+}
+
+#[derive(Clone, Debug)]
+pub struct SvidUpdate {
+    pub svids: HashMap<Id, Svid>,
+}
+
+#[derive(Clone, Debug)]
+pub(crate) struct Api<S> {
+    client: Client<S>,
+}
+
+#[derive(Clone)]
+pub(crate) struct GrpcRecover(ExponentialBackoff);
+
+pub(crate) type Watch<S> = StreamWatch<GrpcRecover, Api<S>>;
+
+// === impl Svid ===
+
+impl TryFrom<api::X509svid> for Svid {
+    type Error = Error;
+    fn try_from(proto: api::X509svid) -> Result<Self, Self::Error> {
+        let cert_der_blocks = asn1::from_der(&proto.x509_svid)?;
+        let (leaf, intermediates) = match cert_der_blocks.split_first() {
+            None => return Err("empty cert chain".into()),
+            Some((leaf_block, intermediates_block)) => {
+                let leaf = DerX509(asn1::to_der(leaf_block)?);
+                let mut intermediates = vec![];
+                for block in intermediates_block.iter() {
+                    let cert_der = asn1::to_der(block)?;
+                    intermediates.push(DerX509(cert_der));
+                }
+                (leaf, intermediates)
+            }
+        };
+
+        let spiffe_id = Id::parse_uri(&proto.spiffe_id)?;
+
+        Ok(Svid {
+            spiffe_id,
+            leaf: leaf.clone(),
```

Comment:
Do we need the clone here? Can we just move the leaf into the Svid struct? Don't think we're going to be sharing the leaf anywhere else outside of this block.